src/libraries/Microsoft.PowerFx.Core/Functions/UserDefinedFunction.cs

```diff
@@ -0,0 +1,229 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Xml.Linq;
+using Microsoft.PowerFx.Core.App;
+using Microsoft.PowerFx.Core.App.ErrorContainers;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Binding.BindInfo;
+using Microsoft.PowerFx.Core.Entities;
+using Microsoft.PowerFx.Core.Errors;
+using Microsoft.PowerFx.Core.Functions;
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Localization;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Types;
+using Microsoft.PowerFx.Core.Utils;
+using Microsoft.PowerFx.Syntax;
+using Microsoft.PowerFx.Types;
+using static Microsoft.PowerFx.Core.Localization.TexlStrings;
+
+namespace Microsoft.PowerFx.Core.Functions
+{
+    internal class UserDefinedFunction : TexlFunction
+    {
+        private readonly bool _isImperative;
+        private readonly IdentToken _returnTypeToken;
+        private readonly ISet<UDFArg> _args;
+
+        public TexlNode UdfBody { get; }
+
+        public override bool IsSelfContained => !_isImperative;
+
+        public UserDefinedFunction(string name, IdentToken returnType, TexlNode body, bool isImperative, ISet<UDFArg> args)
+        : base(DPath.Root, name, name, SG("Custom func " + name), FunctionCategories.UserDefined, returnType.GetFormulaType()._type, 0, args.Count, args.Count, args.Select(a => a.VarType.GetFormulaType()._type).ToArray())
+        {
+            this._returnTypeToken = returnType;
+            this._args = args;
+            this._isImperative = isImperative;
+        }
+
+        public TexlBinding Bind(INameResolver nameResolver, IBinderGlue documentBinderGlue, IUserDefinitionSemanticsHandler userDefinitionSemanticsHandler = null)
+        {
+            if (nameResolver is null)
+            {
+                throw new ArgumentNullException(nameof(nameResolver));
+            }
+
+            if (documentBinderGlue is null)
+            {
+                throw new ArgumentNullException(nameof(documentBinderGlue));
+            }
+
+            var parametersRecordType = CheckParameters(out var errors);
+            var bindingConfig = new BindingConfig(this._isImperative);
+            var binding = TexlBinding.Run(documentBinderGlue, UdfBody, UserDefinitionsNameResolver.Merge(nameResolver, parametersRecordType), bindingConfig);
```

Comment:
How is it possible to perform type checking on the body unless we already know the type of all the UDFs in this batch? Shouldn't the args and return type checks happen first, then a new name resolver is created with the correct signatures, then that new name resolver would be used to type check the body?

Reply:
We already know the type of each UDF but that type may not match with the body or it may be a unknown type. There may be cases where a() depends b() and type of b() is unknown, in this  case dataflow on the host must re-bind all the dependencies of b().

Reply:
No, this is unacceptable. We will bind all UDFs together as a batch and the host will never need to rebind or care about the order in which the UDFs are bound. We have all the UDF signatures already after we parse. We need to type check the signatures independently from the body, make a new name resolver with just the signatures, then bind the bodies using that name resolver.

Reply:
Discussed offline with Bryan and made some changes as per the discussion.